### ISSUE TITLE
Is it normal for the contrastive loss to reach the flat area soon?

### ISSUE THREAD

**Issue by CuberrChen** [maple-research-lab/AdCo#14]
Hello, this is an interesting work. But when I directly apply AdCo to downstream tasks, such as the generator of negative examples, the contrastive loss item will not decline soon(just half of total epochs). 

The curve will look like this:

![图片](https://user-images.githubusercontent.com/44107003/164707359-db154449-5187-4f5d-8fbb-76dd11f3345e.png)

Do you think this is normal?  Or can you give me some advice?

I'm not particularly clear about the code of your memory update, such as `torch.mean`  in ` g = torch.einsum('cn,nk->ck',[q.T,p_qd])/logits.shape[0] - torch.mul(torch.mean(torch.mul(p_qd,l_neg),dim=0),d_norm).`,  so I deduced the formula myself according to [issue](https://github.com/maple-research-lab/AdCo/issues/8) and rewritten the code of gradient manual update just by vector dot product as the formula in [issue](https://github.com/maple-research-lab/AdCo/issues/8).

**Comment by CuberrChen** [comment]
my code as follow:
`
                    logits = logits / memor_temper
                    p_qd = nn.functional.softmax(logits, dim=1)[:, 1:].unsqueeze(-1) # num_queries x num_negative
                    # anchor_feat: num_queries x rep_dim
                    # d/d_norm shape: num_queries x num_negative x rep_dim
                    grad = p_qd * anchor_feat.unsqueeze(1) - p_qd * anchor_feat.unsqueeze(1) * d * d_norm
                    norm_d = torch.norm(d, dim=-1).unsqueeze(-1)
                    if (norm_d != 0).all():
                        try:
                            grad = -torch.div(grad, norm_d) / (num_queries * memor_temper)
                    ### update

`

I don't know if it's the wrong implementation, because I don't know much about it. I hope you can help me point it out.

I am looking forward to your reply soon.

**Comment by wang3702** [comment]
Here we do this complicated implementation to save time for running experiments. For your case, I think you can simply change the updating rule by applying a new optimizer with maximizing the contrastive loss. That would be more reliable. 
Also, your contrastive loss range is a little strange compared to my setting, which may because of the number of your negative examples.

**Comment by CuberrChen** [comment]
thanks !!! i will have a try with using a new optimizer.
